File: lib/rack_resize/processors/bun_image.rb
require "open3"
require "json"
require "tempfile"

class RackResize::Processors::BunImage
  SCRIPT = File.expand_path("bun_image.js", __dir__)

  def resize(source_file:, target_width:, target_height:, target_file: nil, fit: nil, format: nil, quality: nil, bg_color: nil)
    quality ||= RackResize.config.default_quality
    cover = (fit == 'cover' || fit == 'crop') && target_width && target_height

    # ponytail: Bun.Image has no crop/extract yet, so `cover` falls back to `fill`
    # (stretches to exact dims). Upgrade when Bun ships extract() or fit:"cover".
    bun_fit = cover ? 'fill' : 'inside'

    payload = {
      source:  source_file.to_s,
      target:  target_file&.to_s,
      width:   target_width&.to_i,
      height:  target_height&.to_i,
      format:  format && (format.to_s.downcase == 'jpg' ? 'jpeg' : format.to_s.downcase),
      quality: quality&.to_i,
      fit:     bun_fit,
    }

    if target_file
      run_bun(payload)
      nil
    else
      out, err, status = Open3.capture3("bun", SCRIPT, JSON.generate(payload))
      raise "bun_image failed: #{err}" unless status.success?
      out
    end
  end

  private

  def run_bun(payload)
    _, err, status = Open3.capture3("bun", SCRIPT, JSON.generate(payload))
    raise "bun_image failed: #{err}" unless status.success?
  end
end

File: lib/rack_resize/processors/bun_image.js
const opts = JSON.parse(Bun.argv[2]);
const { source, target, width, height, format, quality, fit } = opts;

let img = Bun.file(source).image();

if (width && height) {
  img = img.resize(width, height, { fit: fit === "fill" ? "fill" : "inside" });
} else if (width) {
  img = img.resize(width);
} else if (height) {
  // Bun.Image lacks height-only resize; derive width from source aspect ratio.
  const meta = await img.metadata();
  const w = Math.max(1, Math.round((meta.width * height) / meta.height));
  img = new Bun.Image(source).resize(w, height, { fit: "inside" });
}

const q = quality ?? 85;
const outFmt = (format || source.split(".").pop()).toLowerCase();

switch (outFmt) {
  case "jpg":
  case "jpeg": img = img.jpeg({ quality: q }); break;
  case "png":  img = img.png(); break;
  case "webp": img = img.webp({ quality: q }); break;
  case "avif": img = img.avif({ quality: q }); break;
  case "heic": img = img.heic({ quality: q }); break;
  default: throw new Error(`unsupported format: ${outFmt}`);
}

if (target) {
  await img.write(target);
} else {
  process.stdout.write(await img.bytes());
}
